javalib/src/main/scala/java/time/Duration.scala

```diff
@@ -0,0 +1,234 @@
+
+  def minusHours(hours: Long): Duration =
+    plus(BigInt(3600) * hours * -1000000000)
+
+  def minusMinutes(minutes: Long): Duration =
+    plus(BigInt(60) * minutes * -1000000000)
+
+  def minusSeconds(seconds: Long): Duration =
+    plus(BigInt(seconds) * -1000000000)
+
+  def minusMillis(millis: Long): Duration = plus(BigInt(millis) * -1000000)
+
+  def minusNanos(nanos: Long): Duration = plus(BigInt(nanos) * -1)
+
+  def multipliedBy(multiplicand: Long): Duration =
+    Duration(totalNanos * multiplicand)
+
+  def dividedBy(divisor: Long): Duration = Duration(totalNanos / divisor)
+
+  def negated(): Duration = Duration(totalNanos * -1)
+
+  def abs(): Duration = if (isNegative) negated else this
+
+  def addTo(temporal: Temporal) =
+    temporal.plus(seconds, ChronoUnit.SECONDS).plus(nanos, ChronoUnit.NANOS)
+
+  def subtractFrom(temporal: Temporal) =
+    temporal.minus(seconds, ChronoUnit.SECONDS).minus(nanos, ChronoUnit.NANOS)
+
+  def toDays(): Long = seconds / 86400
+
+  def toHours(): Long = seconds / 3600
+
+  def toMinutes(): Long = seconds / 60
+
+  def toMillis(): Long = {
+    val (quot, rem) = totalNanos /% 1000000
+    val millis = if (rem < 0) quot - 1 else quot
+    // Reference implementation throws if millis < -9223372036854775000!
+    if (millis >= -9223372036854775000L && millis <= Long.MaxValue)
+      millis.toLong
+    else
+      throw new ArithmeticException("Long overflow")
+  }
+
+  def toNanos(): Long = {
+    // Reference implementation throws if nanos < -9223372036000000000!
+    if (totalNanos >= -9223372036000000000L && totalNanos <= Long.MaxValue)
+      totalNanos.toLong
+    else
+      throw new ArithmeticException("Long overflow")
+  }
+
+  def compareTo(that: Duration) = totalNanos.compare(that.totalNanos)
+
+  override def toString(): String = {
+    val nanos1 = if (seconds >= 0 || nanos == 0) nanos else 1000000000 - nanos
+    val secs = if (seconds >= 0 || nanos == 0) seconds else seconds + 1
+    val mins = secs / 60
+    val secsOfMin = secs % 60
+    val hours = mins / 60
+    val minsOfHour = mins % 60
+    val hourPart = if (hours == 0) "" else hours.toString + "H"
+    val minPart = if (minsOfHour == 0) "" else minsOfHour.toString + "M"
+    val decimals = f"$nanos1%09d".reverse.dropWhile(_ == '0').reverse
+    val decimalPart = if (decimals.isEmpty) "" else "." + decimals
+    val secsPart = secsOfMin match {
+      case 0 if seconds != 0 && nanos == 0 => ""
+      case 0 if seconds < 0                => "-0" + decimalPart + "S"
+      case n                               => n.toString + decimalPart + "S"
+    }
+    "PT" + hourPart + minPart + secsPart
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Duration =>
+      seconds == that.getSeconds && nanos == that.getNano
+
+    case _ => false
+  }
+
+  override def hashCode(): Int = 31 * seconds.hashCode + nanos
+}
+
+object Duration {
+  private def apply(nanos: BigInt): Duration = {
```

Comment:
You should try and go to greater lengths to avoid using big numbers, especially if they're not part of the public API. You could have an `apply` method taking two `Long`s to multiply together, and have it detect overflow other than through big numbers.

Depends on how hard to implement that is, but if you manage, it would be a huge performance gain, not to mention that it would avoid to pull in the big numbers just for `java.time`.